### PR TITLE
Add support for authenticating using an API key

### PR DESCRIPTION
**Pull request by jfahrenkrug**
**Task name**: AppStoreRelease and AppStorePromote

**Description**: This adds support for the App Store Connect API key authentication
method, which is the new recommended method by fastlane (see https://docs.fastlane.tools/app-store-connect-api/)

**Documentation changes required:** (Y/N) Y

**Added unit tests:** (Y/N) Y

**Attached related issue:** (Y/N) Y https://github.com/microsoft/app-store-vsts-extension/issues/168

**Checklist**:
- [X] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/blob/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

This is my first PR for an Azure DevOps task. I have uploaded and tested the changes on an actual server and they work. The changes line up with fastlane's documentation.

### CODE REVIEW

**Comment by trajano** [comment]
Can you provide a link on how to create the API key as well?

**Comment by jfahrenkrug** [comment]
> Can you provide a link on how to create the API key as well?

@trajano Sure, here you go: https://docs.fastlane.tools/app-store-connect-api/#creating-an-app-store-connect-api-key

**Comment by PeterStaev** [comment]
@jfahrenkrug can't wait for this to get approved and merged! Do you think you can also add an option to the service end point so we can reuse the authentication for multiple apps? From what I read here: https://docs.microsoft.com/en-us/azure/devops/extend/develop/auth-schemes?view=azure-devops#token-based-authentication and looking at the current authentication scheme that is available here:https://github.com/microsoft/app-store-vsts-extension/blob/0103a2b5f4a01bb22644f506bdd250d59b4fcd62/app-store-vsts-extension.json#L115-L163 should not be hard to add.

**Comment by jfahrenkrug** [comment]
@PeterStaev Thank you! I had begun to implement it with the end point but ran into issues with the schema. Thank you for showing me where that schema lives. I'll make sure to add it there too.

**Comment by jfahrenkrug** [comment]
@PeterStaev Added support for Service Endpoints in https://github.com/microsoft/app-store-vsts-extension/pull/189/commits/9a6c3b657de449563c552d6274bc0fe83cef58de

**Comment by jfahrenkrug** [comment]
@egor-bryzgalov @EzzhevNikita @alexander-smolyakov I believe this is ready for a review. All the tests pass and I've installed and tested it on an actual Azure DevOps server. Thanks!

**Comment by phalecsjagboro** [comment]
Thanks for attempting to fix this. Token expiry is slightly unbearable at the moment. Great work.

**Comment by anatolybolshakov** [comment]
@jfahrenkrug thank you for contribution, this is great to have! We are going to review it in the near time

**Comment by jfahrenkrug** [comment]
> @jfahrenkrug thank you for contribution, this is great to have! We are going to review it in the near time

That's great news, thanks @anatolybolshakov. Let me know if there's anything I can do to help.

**Comment by trajano** [comment]
Is there a way of using this on Azure DevOps like installing it as a custom market-place extension so we can help verify?

**Comment by jfahrenkrug** [comment]
Yes, it’s described here: https://github.com/microsoft/app-store-vsts-extension/blob/master/CONTRIBUTING.md#testing-task-changes

**Comment by jfahrenkrug** [comment]
@anatolybolshakov Hi Anatoly, I just wanted to check if I can do anything to move this along, I'd love to see this getting merged soon.

File: Tasks/app-store-promote/task.loc.json
Comment (anatolybolshakov) [inline Tasks/app-store-promote/task.loc.json:18]: Could you please bump task versions according to [steps](https://github.com/microsoft/app-store-vsts-extension/blob/master/docs/taskversionbumping.md)?

**Comment by anatolybolshakov** [comment]
@jfahrenkrug thanks for contribution here! Could you please take a look at the comment?

File: Tasks/app-store-promote/task.loc.json
Comment (jfahrenkrug) [reply Tasks/app-store-promote/task.loc.json]: @anatolybolshakov Yes, I just bumped it to `180.0`.

**Comment by jfahrenkrug** [comment]
> @jfahrenkrug thanks for contribution here! Could you please take a look at the comment?

My pleasure, I just pushed the version bump. Let me know if there's anything else I can help with to get this over the finish line.

**Comment by 1Cor125** [comment]
Hoping to see this one get merged soon. Great work.

**Comment by velocitysystems** [comment]
We're hoping to see this merged too. Good work @jfahrenkrug!

**Comment by anatolybolshakov** [comment]
Thanks @jfahrenkrug - we are planning to test it also, it could take some time - we will let you know once it's ready and we are good to merge it.

**Comment by jfahrenkrug** [comment]
> Thanks @jfahrenkrug - we are planning to test it also, it could take some time - we will let you know once it's ready and we are good to merge it.

Thank you for testing it and for the update.

**Comment by jfahrenkrug** [comment]
@anatolybolshakov Hi Anatoly, just checking in if there's any movement on this. Thanks.

**Review by PeterStaev — CHANGES_REQUESTED** [review]
Hey @jfahrenkrug , because of another strange problem I was experiencing I decided to try your version of the task and found some problems. Would be great if you can adjust :)

Note: I have listed the changes only to one of the tasks, but similar changes would be needed to the other one that is using the API key implementation.

File: Tasks/app-store-promote/task.json
Comment (PeterStaev) [inline Tasks/app-store-promote/task.json:69]: @jfahrenkrug Forgot this one: So instead of using a custom key best to use the built-in `apitoken` as otherwise the UI adds a default API Token field that is now not used:
image

I think you should change the name to `apitoken` adjust the code to take from this value and leave the label/description as is. This way it will have correct description on the UI (not the default API Token).

File: Tasks/app-store-release/app-store-release.ts
Comment (jfahrenkrug) [reply Tasks/app-store-release/app-store-release.ts:138]: Hi Peter, thanks for the feedback. I did set `is_key_content_base64` on purpose because otherwise the content of the key string was being corrupted. See this commit: https://github.com/microsoft/app-store-vsts-extension/pull/189/commits/a98d239d12d4159cd38bd5322990ee010f2c4b95

Comment (PeterStaev) [reply Tasks/app-store-release/app-store-release.ts:138]: Strange, I had no problems pasting the multiline content, but I used a service connection. So may be the problem is when you enter it in the build definition. Ok, I guess this should be fine then (sorry, I missed that this needs to be base64 of the whole P8 file, I thought it was just the base64 that is inside the file and it was failing for me when I used it like this)

File: Tasks/app-store-promote/app-store-promote.ts
Comment (jfahrenkrug) [reply Tasks/app-store-promote/app-store-promote.ts:76]: Thanks, I'll fix it.

File: Tasks/app-store-promote/app-store-promote.ts
Comment (jfahrenkrug) [reply Tasks/app-store-promote/app-store-promote.ts:55]: Great idea, I'll fix that, thanks.

File: Tasks/app-store-promote/task.json
Comment (PeterStaev) [reply Tasks/app-store-promote/task.json:69]: Simply change the `name` from the current `apiKeyContent` to `apitoken`. This should reuse the built-in field that gets added. Of course you will have to change the usages of `apiKeyContent` and use `apitoken` instead.

File: Tasks/app-store-release/app-store-release.ts
Comment (jfahrenkrug) [reply Tasks/app-store-release/app-store-release.ts:138]: No problem :)

File: Tasks/app-store-promote/task.json
Comment (jfahrenkrug) [reply Tasks/app-store-promote/task.json:69]: I see, I'll make that change. Btw, are you a maintainer on this project?

Comment (PeterStaev) [reply Tasks/app-store-promote/task.json:69]: No, I'm not :(

**Comment by ED220Webfactory** [comment]
Would be great if this could be merged somewhat soonish, since Apple is starting to enforce 2FA this february and the current deployment options would no longer work due to that

**Comment by joke1410** [comment]
Hey @jfahrenkrug! I see you've done some great work here. Since Apple has already enforced 2FA I have a question whether you do plan to merge this PR? I think many would be grateful :)

**Comment by jfahrenkrug** [comment]
> Hey @jfahrenkrug! I see you've done some great work here. Since Apple has already enforced 2FA I have a question whether you do plan to merge this PR? I think many would be grateful :)

Hi, thank you. I'd love for this to get merged, but it's not up to me: I just wrote the code, but Microsoft has to approve and merge it. It's honestly a little disappointing that it's taking so long... they are probably very busy, but it doesn't make me want to contribute to the project again, to be honest.

**Comment by ekrapfl** [comment]
@anatolybolshakov are you a maintainer who could merge this PR?  It would be really nice to get this out there ASAP, now that Apple is enforcing 2FA on everything.

**Comment by Robroots** [comment]
This became a blocker for us as well.

**Comment by StormPooper** [comment]
Looks like the delay for approving this has caused some merge conflicts to appear, @jfahrenkrug will need to update it before one of the maintainers can merge it in.

**Comment by egor-bryzgalov** [comment]
We've faced some issues while rebasing this branch.. We've restored it and updated it to keep up with the latest changes in the main repository. We're going to test these changes to be able to merge them.

**Comment by jfahrenkrug** [comment]
> We've faced some issues while rebasing this branch.. We've restored it and updated it to keep up with the latest changes in the main repository. We're going to test these changes to be able to merge them.

Thanks so much for moving this along @egor-bryzgalov, I really appreciate it. If you notice anything in your code review or testing that needs to be changed, let me know. I'm happy to help getting this over the finish line.

File: Tasks/app-store-promote/task.json
Comment (jfahrenkrug) [reply Tasks/app-store-promote/task.json]: @egor-bryzgalov What are your thoughts on this? Should we change this as suggested by @PeterStaev?

File: Tasks/app-store-promote/app-store-promote.ts
Comment (jfahrenkrug) [reply Tasks/app-store-promote/app-store-promote.ts]: Fixed in https://github.com/microsoft/app-store-vsts-extension/pull/189/commits/b5c7082b840ec7a974f0c1a05d7ea97bf1e27a17

File: Tasks/app-store-promote/task.json
Comment (DaniilShmelev) [reply Tasks/app-store-promote/task.json]: @PeterStaev Could you please clarify why you want this change to be made? I don't see any UI fields being added

image

Comment (PeterStaev) [reply Tasks/app-store-promote/task.json]: @DaniilShmelev , that's because you are looking at it in GitHub actions 😄 In latest version of Azure DevOps Server 2020, it shows an API Token field that is unused:
image

Comment (DaniilShmelev) [reply Tasks/app-store-promote/task.json]: Thanks for the great point @PeterStaev. Yeah, we've checked it, this feature is present in the service connection editor.
@jfahrenkrug, I think it makes sense to apply the suggested change.

Comment (jfahrenkrug) [reply Tasks/app-store-promote/task.json]: Thanks @PeterStaev and @DaniilShmelev, I've made the change in this commit: https://github.com/microsoft/app-store-vsts-extension/pull/189/commits/2e1cf19b3913ece57a4c8e12348848998f7dacf0 I haven't gotten a chance to actually verify it on an Azure DevOps instance, though. If one of you gets a chance to test it out, that would be great. When you do, could you report back WHERE Azure displays the "Key Content" field now? Before this change, the fields were as follows:
```
[App Store Connect API Key ID]
[App Store Connect API Issuer ID]
[App Store Connect API Key Content (base64-encoded)]
[App Store Connect API In House (Enterprise)]
[API Token]
```

With this change from `apiKeyContent` to `apitoken`, is the new order now (A) or (B)?

## Option A:
```
[App Store Connect API Key ID]
[App Store Connect API Issuer ID]
[App Store Connect API In House (Enterprise)]
[App Store Connect API Key Content (base64-encoded)]
```

## Option B:
```
[App Store Connect API Key ID]
[App Store Connect API Issuer ID]
[App Store Connect API Key Content (base64-encoded)]
[App Store Connect API In House (Enterprise)]
```

I'd prefer option B, because it groups all the important, required fields together on top, instead of listing one of the most important fields ("Key Content") at the end, after the optional "Enterprise" field.

Comment (PeterStaev) [reply Tasks/app-store-promote/task.json]: I just tested and it is in the "Option B" place. Thanks for doing these changes!
image

Comment (jfahrenkrug) [reply Tasks/app-store-promote/task.json]: 🥳 That's great, thanks for testing it Peter.

**Comment by DaniilShmelev** [comment]
I've set `--precheck_include_in_app_purchases` to  `false` since fastlane doesn't support prechecking in-app purchases when using API key:

image

**Comment by DaniilShmelev** [comment]
Run `make.js build` and `make.js test` manually on:
✔ windows
✔ centos 7
✔ macos 10.15

✔ Run extension build pipeline

Tested updated tasks:

AppStorePromote:
✔ service connection basic self-hosted
✔ no service connection basic self-hosted
✔ no service connection api key self-hosted
✔ service connection basic ms-hosted
✔ no service connection basic ms-hosted
✔ no service connection api key ms-hosted
❌ service connection api key local AzDO instance - we found some issues with secret masking in service endpoint. We're going to explore our options here

AppStoreRelease:
✔ service connection basic self-hosted
✔ no service connection basic self-hosted
✔ no service connection api key self-hosted
✔ service connection basic ms-hosted
✔ no service connection basic ms-hosted
✔ no service connection api key ms-hosted

File: Tasks/AppStorePromote/Tests/L0.ts
Comment (DaniilShmelev) [inline Tasks/AppStorePromote/Tests/L0.ts:149]: These cleanup lines don't get executed if any assertions above fail. This leaves the `api_keyD383SF739.json` file in the repository. @jfahrenkrug could you please take a look at this issue?

Comment (jfahrenkrug) [reply Tasks/AppStorePromote/Tests/L0.ts:149]: @DaniilShmelev Great catch, thanks, looking at it now.

Comment (jfahrenkrug) [reply Tasks/AppStorePromote/Tests/L0.ts]: @DaniilShmelev fixed in https://github.com/microsoft/app-store-vsts-extension/pull/189/commits/45ce07305b71d8ed95a3634a95253ed257c5db56

**Comment by DaniilShmelev** [comment]
AppStorePromote:
✔ service connection api key
AppStoreRelease:
✔ service connection api key release to testflight
✔ service connection api key release to production

File: Tasks/AppStorePromote/app-store-promote.ts
Comment (egor-bryzgalov) [inline Tasks/AppStorePromote/app-store-promote.ts:55]: I would suggest changing the variable name to `apiKeyFile` or `apiKeyFilePath` since it's actually a path.

File: Tasks/AppStorePromote/app-store-promote.ts
Comment (egor-bryzgalov) [inline Tasks/AppStorePromote/app-store-promote.ts:74]: I'd suggest to use `Agent.TempDirectory` or at least swap these variables since `Agent.TempDirectory` is cleaned up after each job and it seems to be safer.

**Review by DaniilShmelev — APPROVED** [review]
Take a look at Egor's notes, but other that that LGTM. We need at least 2 approvals to merge these changes, so let's wait for someone else to approve them.

File: Tasks/AppStorePromote/app-store-promote.ts
Comment (jfahrenkrug) [reply Tasks/AppStorePromote/app-store-promote.ts:55]: Great point, I'll fix it.

File: Tasks/AppStorePromote/app-store-promote.ts
Comment (jfahrenkrug) [reply Tasks/AppStorePromote/app-store-promote.ts:74]: That's a good idea, I'll fix it now.

File: Tasks/AppStorePromote/app-store-promote.ts
Comment (jfahrenkrug) [reply Tasks/AppStorePromote/app-store-promote.ts]: @egor-bryzgalov Fixed in https://github.com/microsoft/app-store-vsts-extension/pull/189/commits/b79c37a1495a08eaf00671b879017554ac815080

File: Tasks/AppStorePromote/app-store-promote.ts
Comment (jfahrenkrug) [reply Tasks/AppStorePromote/app-store-promote.ts]: @egor-bryzgalov  Fixed in https://github.com/microsoft/app-store-vsts-extension/pull/189/commits/b79c37a1495a08eaf00671b879017554ac815080

**Review by anatolybolshakov — APPROVED** [review]

**Comment by anatolybolshakov** [comment]
LGTM, thanks for contribution!

**Comment by jfahrenkrug** [comment]
> LGTM, thanks for contribution!

@anatolybolshakov  Great, thanks so much. What's the next step now that we have two approvals? Who can actually merge this? :)

**Comment by joke1410** [comment]
Hey @DaniilShmelev @anatolybolshakov, is there any chance to have this merged and available quite soon?

**Review by max-zaytsev — APPROVED** [review]

**Comment by DaniilShmelev** [comment]
Thank you for this contribution @jfahrenkrug! We're going to release the updated extension soon

**Comment by jfahrenkrug** [comment]
> Thank you for this contribution @jfahrenkrug! We're going to release the updated extension soon

@DaniilShmelev Thanks so much for reviewing and merging :)

**Pull request merged**
